### ISSUE TITLE
Is it possible to track the exposed variables when using tgpt/DuckAI

### ISSUE THREAD

**Issue by cameronj86** [aandrew-me/tgpt#298]
I assumed that there was no variables being shared but I'm having second thoughts now... I have the following function to interact w/ one-off questions:
```
function meta() {
        tgpt --provider duckduckgo --model meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo  "${*@Q}"
    }
```
Today, I used Meta's model for a simple, one-off question and it either knew my username to determine $HOME or (more concerning) it can see my directories:
![image](https://github.com/user-attachments/assets/b1605077-bc86-4464-83b9-56fc18e8e39e)
This isn't info I shared explicitly in prior convos w/ Meta specifically 🤔

**Comment by cameronj86** [comment]
No specific steps. Unsolicited, the LLM mentioned my $HOME... I almost exclusively use mistral, which makes it even weirder. Here _all_ the queries I've run using DDG's llama model:
![image](https://github.com/user-attachments/assets/06fd3e05-eb04-4fcc-b177-4ad525710571)

**Edit:** On a whim, I ran the same query and received the same response, so try asking llama the following:
`are config files and databases typically stored in /usr/bin ~/.local/bin/ etc or are they normally housed elsewhere?`

**Comment by cameronj86** [comment]
Ummmm fuck.

At the least it has full read access and that isn't limited to $HOME. Here are the results of 
$ `pwd`
`/usr/local/man`
$`q can /usr/local/man/man1 have a combinationa of *.1 man files and *.gz man files? What is the diff between the two?`

```The files you mentioned are manual pages (man pages) for various Unix/Linux command-line tools. Here's a brief description of each:

1. clambc.1: This is the man page for the ClamAV command-line scanner in BC (Buddha ClamAV) mode.
2. clamconf.1: This is the man page for the ClamAV configuration file.
3. clamdscan.1: This is the man page for the ClamAV on-demand scanner.
4. clamdtop.1: This is the man page for the ClamAV real-time monitoring tool.
5. clamscan.1: This is the man page for the ClamAV command-line scanner.
6. clamsubmit.1: This is the man page for the ClamAV virus database submission utility.
7. freshclam.1: This is the man page for the ClamAV virus database updater.
8. less.1: This is the man page for the pager utility less.
9. lessecho.1: This is the man page for the lessecho utility, which is used to display text using less.
10. lesskey.1: This is the man page for the lesskey utility, which is used to customize the behavior of less.
1. sigtool.1: This is the man page for the ClamAV signature tool.
12. visidata.1.gz: This is the man page for the terminal data visualization tool VisiData.
13. vd.1.gz: This is the man page for the VisiData command-line interface.
The difference between the two sets of man pages is that the first set (clam*) is related to the ClamAV antivirus software, while the second set (less*, lessecho*, lesskey*, sigtool*, visidata*, vd*) is related to various command-line util
ities for Unix/Linux systems.
```
That's everything in my `/usr/local/man/man1`

I'm trying to figure out how to lock down the access via `firejail` ASAP.  AppArmor too if anyone else has any ideas. Fuck me

**Comment by aandrew-me** [comment]
No, its not shared with bot

**Comment by cameronj86** [comment]
Let me be more concise as well:

1. How is the LLM referencing the contents of my directory if it hasn't been explicitly shared in the context window?
2. Why does the tool need/have access to system-wide directories?

**Comment by aandrew-me** [comment]
It doesn't have access to directory contents, unless you explicitly share them.
The tool doesn't need access, it automatically has access because of how permissions work in linux. Use firejail or just don't use it if you don't trust

**Comment by cameronj86** [comment]
> It doesn't have access to directory contents, unless you explicitly share them.

1. I did not share any variables. **The function in my first post shows that I am not using interactive mode, so the LLMs shouldn't be able to create context window outside of an individual prompt even if I wanted them to.**
2. In what workflow would I need to pipe/expose every file I have in `/usr/local/man/` before asking such a simple question?

> Use firejail or just don't use it if you don't trust
I did trust it which is why I loaded a ticket to get env. variables instead of immediately reaching more pessimistic conclusions.

Twice now, I have given examples of more variables being shared than OS Name & Shell Name. I have also given you the query that caused the issue to see if it is something you can replicate. 

I cannot speak to the reason why you cannot/will not acknowledge that more variables are being shared but if anyone else is reading this, can you try recreating either of the queries that I used in the screenshots above to see if you also get similar responses?

**Comment by aandrew-me** [comment]
Also I don't see quotation marks in your queries. It should be like this:
`tgpt --provider duckduckgo --model meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo "are config files and databases typically stored in /usr/bin ~/.local/bin/ etc or are they normally housed elsewhere"`

**Comment by aandrew-me** [comment]
One thing you can do is proxy the requests, you will see exactly what is being sent to the APIs. You can use something like [mitmproxy](https://mitmproxy.org/). Then start `mitmweb`, install the certificate, then set the proxy in tgpt

**Comment by aandrew-me** [comment]
Reply to the second query

```
 tgpt --provider duckduckgo --log /$HOME/logs --model mistralai/Mixtral-8x7B-Instruct-v0.1 "can /usr/local/man/man1 have a combinationa of *.1 man files and *.gz man files? What is the diff between the two?"

 Yes, the directory /usr/local/man/man1 can contain both regular *.1 man files and compressed *.gz man files. Regular man files are stored in plain text
 format, while compressed man files are stored in a compressed format using gzip.
The difference between the two is that regular man files are easier to modify and view with standard text editors, but they take up more disk space compared to comp
ressed man files. On the other hand, compressed man files take up less disk space, but they require additional processing time to decompress and display the
 contents.
When you run the man command to view a man page, the system automatically checks for a compressed version of the man page and decompresses it on the fly if it
 exists. This means that you can view compressed man pages just like regular man pages, without having to manually decompress them first.
Here's an example of how you might see both types of man files in the /usr/local/man/man1 directory:
``bash
/usr/local/man/man1$ ls
example.1.gz
myprogram.1
```
In this example, example.1.gz is a compressed man file, while myprogram.1 is a regular man file. You can view the contents of myprogram.1 with the man
 command like this:
``bash
/usr/local/man/man1$ man myprogram.1
```
And you can view the contents of example.1.gz with the man command like this:
``bash
/usr/local/man/man1$ man example.1.gz
```
In both cases, the system will automatically detect the type of man file and display the contents accordingly.
```

**Comment by aandrew-me** [comment]
> One thing you can do is proxy the requests, you will see exactly what is being sent to the APIs. You can use something like [mitmproxy](https://mitmproxy.org/). Then start `mitmweb`, install the certificate, then set the proxy in tgpt

You can use [httptoolkit ](https://httptoolkit.com/) too

**Comment by cameronj86** [comment]
> **Also I don't see quotation marks in your queries.** It should be like this: `tgpt --provider duckduckgo --model meta-llama/Meta-Llama-3.1-70B-Instruct-Turbo "are config files and databases typically stored in /usr/bin ~/.local/bin/ etc or are they normally housed elsewhere"`

Thanks. I ran it through --debug in firejail and cross-referenced your results and I believe I've figured it out....

I wasn't crazy, just lazy, and it bit me in the ass:
![image](https://github.com/user-attachments/assets/3493b0b7-7213-44fc-b020-a1e05521c3eb)
![image](https://github.com/user-attachments/assets/7ad5aa3d-a9fc-4e4a-8829-531990dc753c)

1. Since I wasn't using quotations, the shell was intercepting and interpreting terminal-specific characters before passing them output onto tgpt. Since I wasn't explicitly typing this in, I didn't realize I was explicitly sharing it by not using quotations.
2. The 'exposing' of variables was infrequent because of the one-off prompts that I use, as well as the separate variables needed (included globs, while being inside of the proper directory, while asking a relevant question)

Definitely felt like I was being gaslit in the moment 😭  I'll close the tix out tomorrow in case you want to throw an "I told you so" in there or assign me some grunt work like working on documentation to atone.
